### PR TITLE
#27 FEATURE: Configure expiration time for passwords

### PR DESCRIPTION
**Pull request by dgrammlich**
Hi!

We implemented the expiration time for passwords for a customer and wanted to show you the solution. Maybe you can give us feedback on our code and / or merge it if you like it.

We are excited to hear from you!

Cheers
Alexander and Daniela

### CODE REVIEW

**Comment by Bentong021** [comment]
Nice

**Comment by Bentong021** [comment]
https://uit.stanford.edu/service/accounts/passwords/quickguide

**Comment by Bentong021** [comment]
https://uit.stanford.edu/newsletter

**Comment by tflidd** [comment]
@LukasReschke @nickvergessen @MorrisJobke any reason we never gave any feedback to this pull request?

File: lib/Db/UserDAO.php [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+
+    /**
+     *  Delete expiration data for one userUid
+     *
+     * @param $userUid
+     */
+    public function deleteUserExpirationData($userUid) {
+        $sql = 'DELETE FROM oc_password_policy_expiration';
+        $sql .= ' WHERE uid="' . $userUid . '"';
```
Comment (nickvergessen) [inline lib/Db/UserDAO.php]: This is a SQL injection.
Must use a parameter at least.

Best would be to use `$this->db->getQueryBuilder()` instead.

Same applies to all other queries.

File: lib/Hook/UserHooks.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
+<?php
+namespace OCA\Password_Policy\Hook;
```
Comment (nickvergessen) [inline lib/Hook/UserHooks.php]: Missing header with Licence etc.
Same for all other files.

File: lib/Db/UserDAO.php [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+     *
+     * @param array $fields
+     *
+     * @return int number of affected rows
+     */
+    public function updateUserExpirationData(array $fields) {
+        $sql = "INSERT INTO oc_password_policy_expiration";
+        $sql .= " (" . implode(',', array_keys($fields)) .  ")";
+        $sql .= ' VALUES ("' . implode('","', $fields) .'")';
+        $sql .= " ON DUPLICATE KEY UPDATE ";
```
Comment (nickvergessen) [inline lib/Db/UserDAO.php]: This does not work on all supported databases.

Either use `$this->db->insertIfNotExist()` and depending on the result do the update or manually select before the update/insert

File: lib/Db/UserDAO.php [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+class UserDAO {
+
+    private $db;
+
+    public function __construct(IDBConnection $db) {
+        $this->db = $db;
+    }
+
+    public function findAllUsers() {
+        $sql = 'SELECT * FROM oc_users LEFT JOIN oc_password_policy_expiration ON oc_users.uid = oc_password_policy_expiration.uid';
```
Comment (nickvergessen) [inline lib/Db/UserDAO.php]: columns need to be escaped with \` or you use the query builder which does it automatically.

File: lib/BackgroundJobs/NotifyUserTask.php [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+            $plaintext_template = new TemplateResponse('password_policy', 'email.usernotify_plaintext', $template_var, 'blank');
+            $plaintext_part = $plaintext_template->render();
+
+            $subject = $this->l10n->t('Your password is about to expire.');
+            $from = Util::getDefaultEmailAddress('register');
+
+            $message = $this->mailer->createMessage();
+            $message->setFrom([$from => $this->defaults->getName()]);
+            $message->setTo([$userEmail => $userId]);
+            $message->setSubject($subject);
```
Comment (nickvergessen) [inline lib/BackgroundJobs/NotifyUserTask.php]: In the meantime setTemplate was added, which comes with nice and easy HTML emails.
Maybe you could also have a look at this.

File: lib/BackgroundJobs/NotifyUserTask.php [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+    public function __construct(PasswordPolicyConfig $config, IMailer $mailer, Defaults $defaults,
+        IUserManager $userManager,
+        IL10N $l10n,
+        IGroupManager $groupManager ) {
+
+        // Run once a day
+        $this->setInterval(24 * 60 * 60);
+
+        $this->config = $config;
+        $this->logger = \OC::$server->getLogger();
```
Comment (nickvergessen) [inline lib/BackgroundJobs/NotifyUserTask.php]: Inject it instead.

File: lib/BackgroundJobs/NotifyUserTask.php [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+        $this->mailer = $mailer;
+        $this->defaults = $defaults;
+        $this->userManager = $userManager;
+        $this->l10n = $l10n;
+        $this->groupManager = $groupManager;
+
+        $connection = \OC::$server->getDatabaseConnection();
+        $this->userDAO = new UserDAO($connection);
```
Comment (nickvergessen) [inline lib/BackgroundJobs/NotifyUserTask.php]: Same with this.

**Comment by wiswedel** [comment]
ping @dgrammlich

**Comment by nickvergessen** [comment]
Well sionce this is almost two years old, I guess it needs some more changes.

Additionally to the above list, emails should use the email template system and database.xml is not recommended anymore.

**Comment by dgrammlich** [comment]
Hi, thanks for your review, @nickvergessen, and sorry for the late feedback. Due to a longer illness at the time of your review I couldn't respond and then just lost track of this issue.

Nevertheless I'm no longer a member of the project and the company. Maybe @alex-boehm can tell more, or you can just close this MR due to obsolete code.

**Comment by jotoeri** [comment]
/me just pushing the button here. 🙈 
Should also be implemented meanwhile in #101

**Pull request closed without merging**